### PR TITLE
Support arrays for Geometry codecs

### LINKED ISSUE #394: Can't bind Boolean[] types
#### Versions

- Driver: 0.9.0.M1
- Database: PostgreSQL 13.1 (Debian 13.1-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit
- Java: openjdk version "11.0.9.1" 2020-11-04
- OS: Microsoft Windows [Version 10.0.19042.867]

#### Current Behavior

The following code shows how I send an `Integer[]` and a `Boolean[]` array through R2DBC to the server:

```java
System.out.println(Arrays.asList(
Flux.from(dbSetup.connectionFactory.create())
    .flatMap(c -> c
        .createStatement("select $1")
        .bind(0, new Integer[] { 0, 1 })
        .execute())
    .flatMap(it -> it.map((r, m) -> r.get(0, Integer[].class)))
    .collectList()
    .block().get(0)));

System.out.println(Arrays.asList(
Flux.from(dbSetup.connectionFactory.create())
    .flatMap(c -> c
        .createStatement("select $1")
        .bind(0, new Boolean[] { false, true })
        .execute())
    .flatMap(it -> it.map((r, m) -> r.get(0, Boolean[].class)))
    .collectList()
    .block().get(0)));
```

The logs show that the `Integer[]` passes (and so do e.g. `Long[]`, `String[]`

```
[0, 1]
```

But `Boolean[]` (or e.g. `LocalDate[]`, `Float[]`, `Double[]`) doesn't work:

```
Exception in thread "main" java.lang.IllegalArgumentException: Cannot encode parameter of type [Ljava.lang.Boolean; ([Ljava.lang.Boolean;@64e43491)
	at io.r2dbc.postgresql.codec.DefaultCodecs.encode(DefaultCodecs.java:217)
	at io.r2dbc.postgresql.ExtendedQueryPostgresqlStatement.bind(ExtendedQueryPostgresqlStatement.java:89)
	at io.r2dbc.postgresql.ExtendedQueryPostgresqlStatement.bind(ExtendedQueryPostgresqlStatement.java:47)
	at org.jooq.testscripts.R2DBC.lambda$3(R2DBC.java:35)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:386)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:120)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1815)
	at reactor.core.publisher.MonoFlatMap$FlatMapInner.onNext(MonoFlatMap.java:249)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1815)
	at reactor.core.publisher.MonoDelayUntil$DelayUntilCoordinator.signal(MonoDelayUntil.java:264)
	at reactor.core.publisher.MonoDelayUntil$DelayUntilTrigger.onComplete(MonoDelayUntil.java:344)
	at reactor.core.publisher.MonoIgnoreElements$IgnoreElementsSubscriber.onComplete(MonoIgnoreElements.java:88)
	at reactor.core.publisher.FluxConcatIterable$ConcatIterableSubscriber.onComplete(FluxConcatIterable.java:121)
	at reactor.core.publisher.MonoIgnoreElements$IgnoreElementsSubscriber.onComplete(MonoIgnoreElements.java:88)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:846)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:608)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:588)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onComplete(FluxFlatMap.java:465)
	at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onComplete(FluxContextWrite.java:126)
	at io.r2dbc.postgresql.util.FluxDiscardOnCancel$FluxDiscardOnCancelSubscriber.onComplete(FluxDiscardOnCancel.java:99)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onComplete(FluxMapFuseable.java:150)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onComplete(FluxMapFuseable.java:150)
	at reactor.core.publisher.FluxWindowPredicate$WindowPredicateMain.checkTerminated(FluxWindowPredicate.java:527)
	at reactor.core.publisher.FluxWindowPredicate$WindowPredicateMain.drainLoop(FluxWindowPredicate.java:475)
	at reactor.core.publisher.FluxWindowPredicate$WindowPredicateMain.drain(FluxWindowPredicate.java:419)
	at reactor.core.publisher.FluxWindowPredicate$WindowPredicateMain.onComplete(FluxWindowPredicate.java:299)
	at reactor.core.publisher.FluxCreate$BaseSink.complete(FluxCreate.java:439)
	at reactor.core.publisher.FluxCreate$BufferAsyncSink.drain(FluxCreate.java:784)
	at reactor.core.publisher.FluxCreate$BufferAsyncSink.complete(FluxCreate.java:732)
	at reactor.core.publisher.FluxCreate$SerializedFluxSink.drainLoop(FluxCreate.java:240)
	at reactor.core.publisher.FluxCreate$SerializedFluxSink.drain(FluxCreate.java:206)
	at reactor.core.publisher.FluxCreate$SerializedFluxSink.complete(FluxCreate.java:197)
	at io.r2dbc.postgresql.client.ReactorNettyClient$Conversation.complete(ReactorNettyClient.java:701)
	at io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.emit(ReactorNettyClient.java:966)
	at io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.onNext(ReactorNettyClient.java:842)
	at io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.onNext(ReactorNettyClient.java:749)
	at reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:118)
	at reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onNext(FluxPeekFuseable.java:854)
	at reactor.core.publisher.FluxMap$MapConditionalSubscriber.onNext(FluxMap.java:220)
	at reactor.core.publisher.FluxMap$MapConditionalSubscriber.onNext(FluxMap.java:220)
	at reactor.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:265)
	at reactor.netty.channel.FluxReceive.onInboundNext(FluxReceive.java:371)
	at reactor.netty.channel.ChannelOperations.onInboundNext(ChannelOperations.java:358)
	at reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:96)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
	Suppressed: java.lang.Exception: #block terminated with an error
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:99)
		at reactor.core.publisher.Mono.block(Mono.java:1703)
		at org.jooq.testscripts.R2DBC.main(R2DBC.java:39)
```

I've noticed https://github.com/pgjdbc/r2dbc-postgresql/issues/391 which is about enum type arrays. Enum type arrays, UDT arrays, and other such things are much harder to support because of the type information that needs to be associated.

I think that supporting `Boolean[]` may appear simpler, and a lower hanging fruit. Or, in other words, arrays of types listed in `R2dbcType` may be different from arrays of user defined types

### PR DESCRIPTION
**Pull request by frankgh**
Make sure that:

- [x] You have read the [contribution guidelines](https://github.com/pgjdbc/r2dbc-postgresql/blob/main/.github/CONTRIBUTING.adoc).
- [x] You have created a feature request first to discuss your contribution intent. Please reference the feature request ticket number in the pull request.
- [x] You use the code formatters provided [here](https://github.com/pgjdbc/r2dbc-postgresql/blob/master/intellij-style.xml) and have them applied to your changes. Don't submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.

 

#### Issue description

resolves #394 (Geometry types)

### CODE REVIEW

**Comment by frankgh** [comment]
@mp911de I added array support for geo types.

File: src/main/java/io/r2dbc/postgresql/codec/CircleCodec.java [hunk 1]
```diff
@@ -50,6 +51,16 @@ ByteBuf doEncodeBinary(Circle value) {
             .writeDouble(value.getRadius());
     }
 
+    @Override
+    public String encodeToText(Circle value) {
+        return String.format("\"%s\"", value);
```
Comment (mp911de) [inline src/main/java/io/r2dbc/postgresql/codec/CircleCodec.java]: `String.format(…)` is rather costly, I'm going to switch instances of `String.format(…)` to string concatenation. `String.format(…)` is fine for exceptions but on hot code paths it creates notable performance drag.

**Comment by mp911de** [comment]
Exciting work, I'm going to do merge the changes likely during next week. This week is busy with releasing all sorts of things.

File: src/main/java/io/r2dbc/postgresql/codec/CircleCodec.java (hunk 1 above)
Comment (frankgh) [reply src/main/java/io/r2dbc/postgresql/codec/CircleCodec.java]: You are right, I pushed https://github.com/pgjdbc/r2dbc-postgresql/pull/403/commits/98c34620171095a8e82a5c846a9236d81ea71470 which replaces `String.format(...)` with string concat in some places.

**Comment by mp911de** [comment]
Sorry, it took so long. This is an exceptionally decent piece of work that qualifies you as a core contributor to this project if you want to keep contributing. Let me know what you think.

**Comment by mp911de** [comment]
Thank you for your contribution. That's merged and polished now.

**Comment by mp911de** [comment]
Kerberos sounds good. We also need to revisit statement/result splitting but that is a topic rooted deeply within reactive streams.

**Pull request closed without merging**